### PR TITLE
Support websocket reconnect on web3-providers-ws.

### LINKED ISSUE #1558: Reconnect on Websocket Drop
Hello, 

I am making a websocket connection to Infura Ropsten
`wss://ropsten.infura.io/ws`

and then making contract event subscription like so
` contract.events.MyEvent()
        .on('data', function(event) {
            console.log('Event Generated', data)
        })
        .on('error', function(error) {
            console.log('Event Error', error)
        })`

Infuras websockets seem to be unstable and drop randomly. Unfortunately the websocket drops silently in the background and my event never receives the error. My app just continues to work and stops receiving event data. 

I can tell that websocket drops by running `setInterval` on
`web3.eth.net.isListening()`

Is there a graceful way to catch the websocket drop and reconnect immediately? If not what is a safe interval to run isListening on?

Thanks!

### PR DESCRIPTION
**Pull request by nujabes403**
## Description

Support websocket reconnect, when connection is closed.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran ```npm run test``` with success and extended the tests if necessary.
- [x] I ran ```npm run build``` and tested the resulting file from ```dist``` folder in a browser.
- [x] I have tested my code on the live network.

#fix #1558 #1852

### CODE REVIEW

**Comment by nujabes403** [comment]
Please check my request @frozeman @nivida 🙂

File: packages/web3-providers-ws/src/index.js [hunk 1]
```diff
@@ -76,6 +76,7 @@ var WebsocketProvider = function WebsocketProvider(url, options)  {
     }
 
     this.connection = new Ws(url, protocol, undefined, headers);
+    this.reconnect = () => new Ws(url, protocol, undefined, headers);
```
Comment (levino) [inline packages/web3-providers-ws/src/index.js:79]: I would think you must bind the result of `this.reconnect()` to `this.connection`. If not then you are awkwardly relying on some side-effect from the Websockets constructor call. My bet is: This does not work at all. Have you tested it?

My suggestion: Add an automated test for this.

File: packages/web3-providers-ws/src/index.js [hunk 2]
```diff
@@ -263,7 +264,9 @@ WebsocketProvider.prototype.send = function (payload, callback) {
         } else {
             console.error('no error callback');
         }
-        callback(new Error('connection not open'));
+        // try reconnect, when connection is gone
+        this.reconnect();
+        callback(new Error('connection not open. try reconnecting...'));
```
Comment (levino) [inline packages/web3-providers-ws/src/index.js:269]: What is the point of throwing an error when at the same trying to solve it. You should silently retry until you are sure there is no way to recover. Then you should stop trying to reconnect and throw an error. This here does not make sense.

**Review by levino — CHANGES_REQUESTED** [review]
I do not think that this code works. Adding a test that shows that it works would help me in my confidence. Also I think https://github.com/ethereum/web3.js/issues/1852 is not an issue that should be accepted. It is okay to throw an error when the websocket connection fails and let the user handle the problem upstream (by recreating the provider for example)

File: packages/web3-providers-ws/src/index.js (hunk 1 above)
Comment (frozeman) [reply packages/web3-providers-ws/src/index.js:79]: I agree

Comment (nujabes403) [reply packages/web3-providers-ws/src/index.js:79]: Yes `binding` will be good. Thanks :)

File: packages/web3-providers-ws/src/index.js (hunk 2 above)
Comment (nujabes403) [reply packages/web3-providers-ws/src/index.js:269]: I agree

Comment (frozeman) [reply packages/web3-providers-ws/src/index.js:269]: There is a need to inform web3.js and the developer, as subscription are socket bound, if the socket breaks and auto-reconnects, developers will wonder why they never get their subscriptions again, as it was basically canceled without them knowing. Also web3.js has some logic of re-subscribing. Please check if that still works! https://github.com/ethereum/web3.js/blob/1.0/packages/web3-core-subscriptions/src/subscription.js#L278-L289

Comment (nujabes403) [reply packages/web3-providers-ws/src/index.js:269]: @frozeman Thanks! I'll check it :)

**Comment by nivida** [comment]
I've closed this PR because of #1966

**Pull request closed without merging**